otelcol.exporter.otlphttp: new component (#2429)
grafana/agent: docs/sources/flow/reference/components/otelcol.exporter.otlphttp.md

---
aliases:
- /docs/agent/latest/flow/reference/components/otelcol.exporter.otlphttp
title: otelcol.exporter.otlphttp
---

# otelcol.exporter.otlphttp

`otelcol.exporter.otlphttp` accepts telemetry data from other `otelcol`
components and writes them over the network using the OTLP HTTP protocol.

> **NOTE**: `otelcol.exporter.otlphttp` is a wrapper over the upstream
> OpenTelemetry Collector `otlphttp` exporter. Bug reports or feature requests
> will be redirected to the upstream repository, if necessary.

Multiple `otelcol.exporter.otlphttp` components can be specified by giving them
different labels.

## Usage

```river
otelcol.exporter.otlphttp "LABEL" {
  client {
    endpoint = "HOST:PORT"
  }
}
```

## Arguments

`otelcol.exporter.otlphttp` supports the following arguments:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------

> `metrics_endpoint` | `string` | The endpoint to send metrics to | `client.endpoint + "/v1/metrics"` | no
> `logs_endpoint`    | `string` | The endpoint to send logs to    | `client.endpoint + "/v1/logs"`    | no
> `traces_endpoint`  | `string` | The endpoint to send traces to  | `client.endpoint + "/v1/traces"`  | no

Reviewer: ```suggestion
`metrics_endpoint` | `string` | The endpoint to send metrics to. | `client.endpoint + "/v1/metrics"` | no
`logs_endpoint`    | `string` | The endpoint to send logs to.    | `client.endpoint + "/v1/logs"`    | no
`traces_endpoint`  | `string` | The endpoint to send traces to.  | `client.endpoint + "/v1/traces"`  | no
```

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`endpoint`           | `string`      | `host:port` to send telemetry data to. | | yes
`read_buffer_size`   | `string`      | Size of the read buffer the HTTP client uses for reading server responses. | `0` | no
`write_buffer_size`  | `string`      | Size of the write buffer the HTTP client uses for writing requests. | `"512KiB"` | no
`timeout`            | `duration`    | Time to wait before marking a request as failed. | `"30s"` | no
`headers`            | `map(string)` | Additional headers to send with the request. | `{}` | no
`compression`        | `string`      | Compression mechanism to use for requests. | `"gzip"` | no
`max_idle_conns`     | `int`         | Limits the number of idle HTTP connections the client can keep open. | `100` | no
`max_idle_conns_per_host` | `int`    | Limits the number of idle HTTP connections the host can keep open. | `0` | no
`max_conns_per_host` | `int`         | Limits the total (dialing,active, and idle) number of connections per host. | `0` | no

> `idle_conn_timeout`  | `duration`    | Time to wait before an idle connection will close itself. | `"90s"` | no

Reviewer: ```suggestion
`idle_conn_timeout`  | `duration`    | Time to wait before an idle connection closes itself. | `"90s"` | no
```

`auth`               | `capsule(otelcol.Handler)` | Handler from an `otelcol.auth` component to use for authenticating requests. | | no

{{< docs/shared lookup="flow/reference/components/otelcol-compression-field.md" source="agent" >}}

### tls block

> The `tls` block configures TLS settings used for the connection to the HTTP

Reviewer: Do you think we should move these introductory sentences into the `shared` files or are these sentences potentially different for different instantiations? 

I notice that we didn't include them in earlier shared files but maybe we should.

Author: Yeah, I kept them separate so we can explicitly call out the HTTP/gRPC server.

We can rewrite for a more vague statement, but then the two pages would look eerily similar, maybe it's good to leave them so people can scan at a glance.

Reviewer: I like what you did.  This is a good approach.

> OpenTelemetry Collector `otlphttp` exporter. Bug reports or feature requests
> will be redirected to the upstream repository, if necessary.

Multiple `otelcol.exporter.otlphttp` components can be specified by giving them
different labels.

## Usage

```river
otelcol.exporter.otlphttp "LABEL" {
  client {
    endpoint = "HOST:PORT"
  }
}
```

## Arguments

`otelcol.exporter.otlphttp` supports the following arguments:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`metrics_endpoint` | `string` | The endpoint to send metrics to | `client.endpoint + "/v1/metrics"` | no
`logs_endpoint`    | `string` | The endpoint to send logs to    | `client.endpoint + "/v1/logs"`    | no
`traces_endpoint`  | `string` | The endpoint to send traces to  | `client.endpoint + "/v1/traces"`  | no

The default value depends on the `endpoint` field set in the required `client`
block. If set, these arguments override the `client.endpoint` field for the
corresponding signal.

## Blocks

The following blocks are supported inside the definition of
`otelcol.exporter.otlphttp`:

Hierarchy | Block | Description | Required
--------- | ----- | ----------- | --------

> `client`           | [client][] | Configures the HTTP server to send telemetry data to. | yes
> `client > tls`     | [tls][] | Configures TLS for the HTTP client. | no
> `queue`            | [queue][] | Configures batching of data before sending. | no
> `retry`            | [retry][] | Configures retry mechanism for failed requests. | no

Reviewer 2: We deliberately don't use code fonts in the hierarchies because it looks weird. [Example](https://grafana.com/docs/agent/latest/flow/reference/components/discovery.kubernetes/#blocks)

```suggestion
client           | [client][] | Configures the HTTP server to send telemetry data to. | yes
client > tls     | [tls][] | Configures TLS for the HTTP client. | no
queue            | [queue][] | Configures batching of data before sending. | no
retry            | [retry][] | Configures retry mechanism for failed requests. | no
```

controls which compression mechanism to use. Supported strings are:

* `"gzip"`
* `"zlib"`
* `"deflate"`
* `"snappy"`
* `"zstd"`

If `compression` is set to `"none"` or an empty string `""`, no compression is
used.

### tls block

The `tls` block configures TLS settings used for the connection to the HTTP
server.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`ca_file` | `string` | Path to the CA file. | | no
`cert_file` | `string` | Path to the TLS certificate. | | no
`key_file` | `string` | Path to the TLS certificate key. | | no
`min_version` | `string` | Minimum acceptable TLS version for connections. | `"TLS 1.2"` | no
`max_version` | `string` | Maximum acceptable TLS version for connections. | `"TLS 1.3"` | no
`insecure` | `boolean` | Disables TLS when connecting to the HTTP server. | | no
`insecure_skip_verify` | `boolean` | Ignores insecure server TLS certificates. | | no
`server_name` | `string` | Verifies the hostname of server certificates when set. | | no

The `tls` block should always be provided, even if the server doesn't support

> TLS. To disable `tls` for connections to the HTTP server, set the `insecure`

Author: @rfratto I copy/pasted this from the `otlp` exporter. We say that this block should always be provided, but it's not a required one. Should we relax this doc sentence (eg. s/always be/always can/)?

Reviewer 2: Yeah, this is my fault, I think I misspoke originally. 

I think it's probably more accurate to say:

> If the server doesn't support TLS, the `tls` block must be provided with the `insecure` argument set to `true`. 

and drop the recommendation about it being provided, since I don't think you _need_ it for connecting to TLS servers unless the server does client certificate auth.

WDYT?

> argument to `true`.

Reviewer 2: There's no component-specific defaults here, right? Should we add this as a new shared docs snippet so we can deduplicate it from the two exporters?

Author: No, it just reuses the `configtls` package. I'll add both this and the queue/retry blocks into shared docs snippets to trim both exporters down.

more quickly at the expense of increased network traffic.

### retry block

The `retry` block configures how failed requests to the HTTP server are
retried.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`enabled` | `boolean` | Enables retrying failed requests. | `true` | no
`initial_interval` | `duration` | Initial time to wait before retrying a failed request. | `"5s"` | no
`max_interval` | `duration` | Maximum time to wait between retries. | `"30s"` | no
`max_elapsed_time` | `duration` | Maximum amount of time to wait before discarding a failed batch. | `"5m"` | no

When `enabled` is `true`, failed batches are retried after a given interval.
The `initial_interval` argument specifies how long to wait before the first
retry attempt. If requests continue to fail, the time to wait before retrying
increases exponentially. The `max_interval` argument specifies the upper bound
of how long to wait between retries.

If a batch has not sent successfully, it is discarded after the time specified
by `max_elapsed_time` elapses. If `max_elapsed_time` is set to `"0s"`, failed

> requests are retried forever until they succeed.

Reviewer 2: It looks like both of these blocks can be introduced as new docs snippets since neither this nor otelcol.exporter.otlp introduce component-specific defaults here.
